File: src/main/java/com/tms/csp/transforms/ListTransform.java
package com.tms.csp.transforms;

import com.tms.csp.ast.Exp;
import com.tms.csp.util.ConstraintSink;

import javax.annotation.Nonnull;
import java.util.ArrayList;
import java.util.List;

import static com.google.common.base.Preconditions.checkState;

public abstract class ListTransform {

    @Nonnull
    public abstract List<Exp> transform(@Nonnull List<Exp> in);

    public static void transform(List<Exp> listIn, Transformer t, ConstraintSink out) {
        for (Exp e : listIn) {
            Exp s = t.transform(e);
            out.addConstraint(s);
        }
    }

    public static List<Exp> transform(List<Exp> listIn, Transformer... transformer) {
        if (transformer.length == 0) {
            return listIn;
        } else if (transformer.length == 1) {
            return transformList(listIn, transformer[0]);
        } else {
            return multiTransformList(listIn, transformer);
        }
    }

    public static List<Exp> transformList(List<Exp> listIn, Transformer transformer) {
        int elementsChanged = 0;
        List<Exp> listOut = new ArrayList<Exp>();
        for (Exp argIn : listIn) {
            checkState(argIn != null);
            Exp argOut = transformer.transform(argIn);
            if (argOut == null) {
                throw new IllegalArgumentException("transform returned null. Transformer[" + transformer.getClass().getName() + "]  expIn[" + argIn + "]");
            } else if (argOut != argIn) {
                elementsChanged++;
                listOut.add(argOut);
            } else {
                listOut.add(argOut);
            }
        }

        if (elementsChanged > 0) {
            return listOut;
        } else {
            return listIn;
        }

    }

    public static Iterable<Exp> transform2(Iterable<Exp> listIn, Transformer transformer) {
        int elementsChanged = 0;
        List<Exp> listOut = new ArrayList<Exp>();
        for (Exp argIn : listIn) {
            checkState(argIn != null);
            Exp argOut = transformer.transform(argIn);
            if (argOut == null) {
                throw new IllegalArgumentException("transform returned null. Transformer[" + transformer.getClass().getName() + "]  expIn[" + argIn + "]");
            } else if (argOut != argIn) {
                elementsChanged++;
                listOut.add(argOut);
            } else {
                listOut.add(argOut);
            }
        }

        if (elementsChanged > 0) {
            return listOut;
        } else {
            return listIn;
        }

    }

    public static List<Exp> multiTransformList(List<Exp> listIn, Transformer[] transformers) {
        List<Exp> out = listIn;
        for (Transformer transformer : transformers) {
            List<Exp> in = out;
            out = transformList(in, transformer);
        }
        return out;
    }


}

File: src/main/java/com/smartsoft/csp/transforms/ListTransform.kt
package com.smartsoft.csp.transforms

import com.smartsoft.csp.ast.Exp
import com.smartsoft.csp.util.ConstraintSink
import java.util.ArrayList

import com.google.common.base.Preconditions.checkState

abstract class ListTransform {

    abstract fun transform(inExp: List<Exp>): List<Exp>

    companion object {

        fun transform(listIn: List<Exp>, t: Transformer, out: ConstraintSink) {
            for (e in listIn) {
                val s = t.transform(e)
                out.addConstraint(s)
            }
        }

        fun transform(listIn: List<Exp>, vararg transformer: Transformer): List<Exp> {
            return if (transformer.size == 0) {
                listIn
            } else if (transformer.size == 1) {
                transformList(listIn, transformer[0])
            } else {
                multiTransformList(listIn, transformer)
            }
        }

        fun transformList(listIn: List<Exp>, transformer: Transformer): List<Exp> {
            var elementsChanged = 0
            val listOut = ArrayList<Exp>()
            for (argIn in listIn) {
                checkState(argIn != null)
                val argOut = transformer.transform(argIn)
                if (argOut == null) {
                    throw IllegalArgumentException("transform returned null. Transformer[" + transformer.javaClass.name + "]  expIn[" + argIn + "]")
                } else if (argOut !== argIn) {
                    elementsChanged++
                    listOut.add(argOut)
                } else {
                    listOut.add(argOut)
                }
            }

            return if (elementsChanged > 0) {
                listOut
            } else {
                listIn
            }

        }

        fun transform2(listIn: Iterable<Exp>, transformer: Transformer): Iterable<Exp> {
            var elementsChanged = 0
            val listOut = ArrayList<Exp>()
            for (argIn in listIn) {
                checkState(argIn != null)
                val argOut = transformer.transform(argIn)
                if (argOut == null) {
                    throw IllegalArgumentException("transform returned null. Transformer[" + transformer.javaClass.name + "]  expIn[" + argIn + "]")
                } else if (argOut !== argIn) {
                    elementsChanged++
                    listOut.add(argOut)
                } else {
                    listOut.add(argOut)
                }
            }

            return if (elementsChanged > 0) {
                listOut
            } else {
                listIn
            }

        }

        fun multiTransformList(listIn: List<Exp>, transformers: Array<out Transformer>): List<Exp> {
            var out = listIn
            for (transformer in transformers) {
                val inExp = out
                out = transformList(inExp, transformer)
            }
            return out
        }
    }


}
